### PR TITLE
[grpc] ConnectionFactory: returns structured error so that a depending mod does not need to deal with every error defined in ErrorCode

### PR DESCRIPTION
**Pull request by drmingdrmer**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

##### [grpc] ConnectionFactory: returns structured error so that a depending mod does not need to deal with every error defined in ErrorCode
- Remove conversion from ErrorCode to MetaError, which is a potential bug maker:
  It can not find out missing matching branch at compile time.

- Remove MetaNetworkError::CannotConnectNode, it should be a sub error of ConnectionError

- Remove MetaError:TLSConfigurationFailure, it should be a sub error of MetaNetworkError

- Return Infallible if a method does not return any error

- fix: #4151

## Related Issues

### CODE REVIEW

**Review by lichuang — APPROVED** [review]

File: common/meta/grpc/src/grpc_client.rs [hunk 1]
```diff
@@ -87,7 +101,9 @@ pub struct MetaGrpcClient {
 const AUTH_TOKEN_KEY: &str = "auth-token-bin";
 
 impl MetaGrpcClient {
-    pub async fn try_new(conf: &MetaGrpcClientConf) -> Result<MetaGrpcClient> {
+    pub async fn try_new(
+        conf: &MetaGrpcClientConf,
+    ) -> std::result::Result<MetaGrpcClient, Infallible> {
```
Comment (ariesdevil) [inline common/meta/grpc/src/grpc_client.rs:106]: Why not just rename `try_new` to `new` and change the return type from `Result<T, Infallible>` to `T` if a method does not return any error.

File: query/src/catalogs/backends/meta_backend.rs [hunk 2]
```diff
@@ -70,7 +70,10 @@ impl MetaBackend {
         F: Send + Sync + 'static,
         T: Send + Sync + 'static,
     {
-        let cli = self.meta_api_provider.try_get_meta_client().await?;
+        let res: Result<_, std::convert::Infallible> =
+            self.meta_api_provider.try_get_meta_client().await;
+        let cli = res.unwrap();
```
Comment (ariesdevil) [inline query/src/catalogs/backends/meta_backend.rs:75]: If we do the above return type modification, we do not need unwrap here.

File: common/meta/grpc/src/grpc_client.rs (hunk 1 above)
Comment (drmingdrmer) [reply common/meta/grpc/src/grpc_client.rs:106]: I'm not very sure about whether in the future an error will return.
IMHO, it is very likely to be.

File: query/src/catalogs/backends/meta_backend.rs (hunk 2 above)
Comment (drmingdrmer) [reply query/src/catalogs/backends/meta_backend.rs:75]: There are two different ways to make codes clean:
- To minimize the logic involved,
- or to follow a universal code style, e.g. a method that creates an instance always returns a `Result`.

In this case, either to minimize logic or to keep the return type of every `new_xxx()` method a `Result`. Both feel good to me.

It does not bother me. Then I'll take your word, sir:D

Comment (ariesdevil) [reply query/src/catalogs/backends/meta_backend.rs:75]: Convinced by you :)

**Pull request merged**